runtime/major_gc.c

```diff
@@ -273,15 +273,114 @@ Caml_inline void mark_stack_push(struct mark_stack* stk, value block,
 }
 
 
+#ifdef NO_NAKED_POINTERS
+#ifdef NAKED_POINTERS_CHECKER
+
+#ifdef _WIN32
+#include <windows.h>
+
+static inline int safe_load(volatile value* p, value* result)
+{
+  value v;
+  __try {
+    v = *p;
+  }
+  __except(GetExceptionCode() == EXCEPTION_ACCESS_VIOLATION ?
+        EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH) {
+    *result = 0xdeadbeef;
+    return 0;
+  }
+  *result = v;
+  return 1;
+}
+#else
+static inline int safe_load (header_t * addr, /*out*/ header_t * contents)
+{
+  int ok;
+  header_t h;
+  intnat tmp;
+
+  asm volatile(
+      "leaq 1f(%%rip), %[tmp] \n\t"
+      "movq %[tmp], 0(%[handler]) \n\t"
+      "xorl %[ok], %[ok] \n\t"
+      "movq 0(%[addr]), %[h] \n\t"
+      "movl $1, %[ok] \n\t"
+  "1: \n\t"
+      "xorq %[tmp], %[tmp] \n\t"
+      "movq %[tmp], 0(%[handler])"
+      : [tmp] "=&r" (tmp), [ok] "=&r" (ok), [h] "=&r" (h)
+      : [addr] "r" (addr),
+        [handler] "r" (&(Caml_state->checking_pointer_pc)));
+  *contents = h;
+  return ok;
+}
+#endif
+
+#if SIZEOF_LONG == SIZEOF_PTR
+#define VAL_ONE 1ul
+#else
+#define VAL_ONE 1ull
+#endif
+
+static int is_pointer_safe (value v, value *p)
+{
+  header_t h;
+  tag_t t;
+
+  if (! safe_load(&Hd_val(v), &h)) goto on_segfault;
+
+  t = Tag_hd(h);
+  if (t == Infix_tag) {
+    v -= Infix_offset_hd(h);
+    if (! safe_load(&Hd_val(v), &h)) goto on_segfault;
+    t = Tag_hd(h);
+  }
+
+  /* For the pointer to be considered safe, either the given pointer is in heap
+   * or the (out of heap) pointer has a black header and its size is < 2 ** 40
+   * words (128 GB). If not, we report a warning. */
+  if (Is_in_heap (v) || (Is_black_hd(h) && Wosize_hd(h) < (VAL_ONE << 40)))
+    return 1;
+
+  if (!Is_black_hd(h)) {
+    fprintf (stderr, "Out-of-heap pointer at %p of value %p has "
+                    "non-black head (tag=%d)\n", p, (void*)v, t);
+  } else {
+    fprintf (stderr, "Out-of-heap pointer at %p of value %p has "
+                      "suspiciously large size: %lu words\n",
+              p, (void*)v, Wosize_hd(h));
+  }
+  return 0;
+
+ on_segfault:
+  fprintf (stderr, "Out-of-heap pointer at %p of value %p. "
+           "Cannot read head.\n", p, (void*)v);
+  return 0;
+}
+
+#endif
+#endif
+
 void caml_darken (value v, value *p /* not used */)
 {
+  header_t h;
+  tag_t t;
 #ifdef NO_NAKED_POINTERS
   if (Is_block (v) && !Is_young (v)) {
+#ifdef NAKED_POINTERS_CHECKER
+    if (!is_pointer_safe(v,p)) return;
+
+    /* Atoms never need to be marked. */
+    if (Wosize_val (v) == 0)
+      return;
+#endif /* NNP check */
```

Comment:
```suggestion
#endif /* NAKED_POINTERS_CHECKER */
```
using the actual macro name is better for readability.